### PR TITLE
Redo how secrets are generated wrapup

### PR DESCRIPTION
**Pull request by jandubois**

### CODE REVIEW

File: secrets/secrets.go [hunk 1]
```diff
@@ -108,17 +154,13 @@ func (sg *SecretGenerator) GetSecretConfig(c ConfigMapInterface) *v1.ConfigMap {
...
-func (sg *SecretGenerator) GetSecret(s SecretInterface, configMap *v1.ConfigMap) *v1.Secret {
+func (sg *SecretGenerator) getSecret(s secretInterface, configMap *v1.ConfigMap) (*v1.Secret, error) {
 	currentName := configMap.Data[currentSecretName]
 
-	newName := sg.Getenv("KUBE_SECRETS_GENERATION_NAME")
-	if newName == "" {
-		sg.Fatal("KUBE_SECRETS_GENERATION_NAME is missing or empty.")
-		return nil
-	}
+	newName := sg.secretsName
```
Comment (mook-as) [inline secrets/secrets.go]: We don't really need this variable anymore do we?

File: secrets/secrets_test.go [hunk 2]
```diff
@@ -95,14 +79,24 @@ func (m *MockConfigMapInterface) Update(configMap *v1.ConfigMap) (*v1.ConfigMap,
 	return nil, nil
 }
 
+func testingSecretGenerator() SecretGenerator {
+	sg := SecretGenerator{
```
Comment (mook-as) [inline secrets/secrets_test.go]:
```golang
return SecretGenerator{…
```

File: ssl/ssl_test.go [hunk 3]
```diff
@@ -318,98 +248,90 @@ func TestAddHost(t *testing.T) {
...
-			[]interface{}{defaultCA})
+		secrets := &v1.Secret{Data: map[string][]byte{}}
 
+		err := createCert(certInfo, "namespace", "suffix", secrets, certID)
+		if assert.Error(t, err) {
+			assert.Equal(t, "CA "+defaultCA+" not found", err.Error())
+		}
 	})
 
 	t.Run("If CA cert fails to parse, it should log a fatal error", func(t *testing.T) {
```
Comment (mook-as) [inline ssl/ssl_test.go]: The test message is no longer correct (no logs).

File: ssl/ssl_test.go [hunk 4]
```diff
@@ -420,86 +342,77 @@ func TestCreateCert(t *testing.T) {
...
+		err := createCert(certInfo, "namespace", "suffix", secrets, certID)
 
-		origGetEnv := getEnv
-		defer func() {
-			getEnv = origGetEnv
-		}()
-		getEnv = func(string) string {
-			return "2"
+		if assert.NoError(t, err) {
+			assert.NotEqual(t, secrets.Data[certInfo[certID].PrivateKeyName], []byte{})
```
Comment (mook-as) [inline ssl/ssl_test.go]: `NotEmpty`?

File: model/manifest.go [hunk 5]
```diff
@@ -89,25 +90,22 @@ type Manifest struct {
...
-	if err != nil {
-		logFatal(err)
-		return
-	}
-
-	if err = yaml.Unmarshal(manifestFile, &manifest); err != nil {
-		logFatal(err)
-		return
+// GetManifest loads a manifest from file or string
+func GetManifest(input string) (manifest Manifest, err error) {
```
[the remark this answers is not in the document]
Comment (jandubois) [reply model/manifest.go]: > I think not having named return values is clearer,

I generally agree, but in this particular case it simplifies the logic a bit, so I found it justified this time.

[the remark this answers is not in the document]
Comment (jandubois) [reply model/manifest.go]: > I also think that passing the input in as an io.Reader will be easier to deal with;

How does this work for testing the `file not found` error? Not that I really care that much; just tried to keep the original set of test cases. I guess I just move the file opening to the caller...

File: ssl/ssl_test.go [hunk 6]
```diff
@@ -318,98 +248,90 @@ func TestAddHost(t *testing.T) {
...
-	t.Run("If secrets has a private key, return false", func(t *testing.T) {
-		var mockLog MockLog
-		logFatalf = mockLog.Fatalf
+	t.Run("If secrets already has a private key, do nothing", func(t *testing.T) {
+		t.Parallel()
 
-		secrets := &v1.Secret{Data: map[string][]byte{}}
-		secrets.Data["private-key"] = []byte("private-key-data")
-		secrets.Data["certificate-name"] = []byte("certificate-data")
+		certInfo := defaultCertInfo
```
[the remark this answers is not in the document]
Comment (jandubois) [reply ssl/ssl_test.go]: > there is no copy involved here

Such a hateful language! 😞 Makes a copy when I don't want it to (`for ... range`), but doesn't when I do.

File: secrets/secrets.go [hunk 7]
```diff
@@ -27,33 +27,81 @@ const currentSecretName = "current-secrets-name"
...
-	return SecretGenerator{
-		Fatal:  log.Fatal,
-		Getenv: os.Getenv,
+	sg := SecretGenerator{
+		namespace:           os.Getenv("KUBERNETES_NAMESPACE"),
+		serviceDomainSuffix: os.Getenv("KUBE_SERVICE_DOMAIN_SUFFIX"),
+		secretsName:         os.Getenv("KUBE_SECRETS_GENERATION_NAME"),
+		secretsGeneration:   os.Getenv("KUBE_SECRETS_GENERATION_COUNTER"),
+	}
+	// XXX All these settings should be passed from the commandline and not the environment
```
[the remark this answers is not in the document]
Comment (jandubois) [reply secrets/secrets.go]: I've been pushing it closer to the edge already, but I don't want to go all the way to pass them as arguments, and then create an ERB file to generate the call. I'll leave that for the next refactoring...

File: model/manifest.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mook-as) [reply model/manifest.go]: Yeah, exactly.

File: secrets/secrets.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mook-as) [reply secrets/secrets.go]: Right, I just meant to push it into `main.go`. So there's still env vars, but not visible to the rest of the app.

**Pull request merged**